datalad_next/tree.py

```diff
@@ -0,0 +1,1172 @@
+                for depth, parent in enumerate(parents_below_root):
+                    if parent not in visited:
+                        visited.add(parent)
+
+                        yield Node(parent, depth)
+
+                visited.add(node.path)
+                yield node
+
+    def _is_valid_dataset(self, node):
+        return isinstance(node, DatasetNode) and \
+               is_path_relative_to(node.path, self.root) and \
+               not self.exclude_node_func(node) and \
+               not self._ds_exceeds_max_ds_depth(node)
+
+    def _ds_exceeds_max_ds_depth(self, ds_node):
+        return ds_node.ds_depth > self.max_dataset_depth
+
+    def _ds_child_node_exceeds_max_depth(self, ds_node):
+        ds_parent = get_dataset_root_datalad_only(ds_node.path)
+        if ds_parent is None:
+            return True  # it's not a dataset child, we exclude it
+
+        ds_parent_depth = self.path_depth(ds_parent)
+        if not self._is_valid_dataset(Node(ds_parent, ds_parent_depth)):
+            return True  # also exclude
+
+        # check directory depth relative to the dataset parent
+        rel_depth = ds_node.depth - ds_parent_depth
+        assert rel_depth >= 0, "relative depth from parent cannot be < 0 " \
+                               f"(path: '{ds_node.path}', parent: '{ds_parent}')"
+        return rel_depth > self.max_depth
+
+    def _is_parent_of_ds(self, node):
+        if isinstance(node, FileNode):
+            return False  # files can't be parents
+
+        if self._next_ds is None:
+            return False  # no more datasets, can't be a parent
+
+        if self._next_ds.path == node.path:
+            # we hit a dataset or the parent of a dataset
+            self._advance_ds_generator()
+            return True
+
+        return False
+
+
+class _TreeNode:
+    """Base class for a directory or file represented as a single tree node
+    and printed as single line of the 'tree' output."""
+    TYPE = None  # needed for command result dict
+
+    def __init__(self, path: Path, depth: int,
+                 exception: CapturedException = None):
+        """
+        Parameters
+        ----------
+        path: Path
+            Path of the tree node
+        depth: int
+            Directory depth of the node within its tree
+        exception: CapturedException
+            Exception that may have occurred at validation/creation
+        """
+        self.path = path
+        self.depth = depth
+        self.exception = exception
+
+    def __eq__(self, other):
+        return self.path == other.path
+
+    def __hash__(self):
+        return hash(str(self.path))
+
+    @property
+    def tree_root(self) -> Path:
+        """Calculate tree root path from node path and depth"""
+        parents = self.parents
+        return parents[0] if parents \
+            else self.path  # we are the root
+
+    @property
+    def parents(self) -> list[Path]:
```

Comment:
> Btw, Appveyor build on Ubuntu and MacOS is failing. Seems to be a problem with loading the command interface. I can run my tests locally just fine though. Maybe Python version issue?

It is caused by the type annotation in Python 3.8:

```
$ python
Python 3.8.10 (default, Mar 15 2022, 12:22:08) 
[GCC 9.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from datalad_next.tree import TreeCommand
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/appveyor/projects/datalad-next/datalad_next/tree.py", line 919, in <module>
    class _TreeNode:
  File "/home/appveyor/projects/datalad-next/datalad_next/tree.py", line 954, in _TreeNode
    def parents(self) -> list[Path]:
TypeError: 'type' object is not subscriptable
>>> 

```

```suggestion
    # More accurate annotation only from PY3.9 onwards
    # def parents(self) -> list[Path]:
    def parents(self) -> list:
```

Reply:
ATTN: I will commit this change to see how the test behave afterwards.

Reply:
Thank you for troubleshooting and pushing the changes!